core/src/main/scala/org/bykn/bosatsu/Normalization.scala

```diff
@@ -0,0 +1,666 @@
+            { (arg: NormalExpression, acc: PatternEnv) =>
+              arg match {
+                case NormalExpression.Struct(_, args) =>
+                  processArgs(args, acc)
+                case _ =>
+                  NotProvable
+              }
+            } 
+
+          case Some(idx) =>
+            // we don't check if idx < 0, because if we compiled, it can't be
+            val result = { (arg: NormalExpression, acc: PatternEnv) =>
+              arg match {
+                case NormalExpression.Struct(enumId, args) =>
+                  if (enumId == idx) processArgs(args, acc)
+                  else NoMatch
+                case _ =>
+                  NotProvable
+              }
+            }
+            result
+        }
+  }
+
+  def findMatch(m: NormalExpression.Match) =
+    m.branches.collectFirst(Function.unlift( { case (pat, result) =>
+      maybeBind(pat).apply(m.arg, Map()) match {
+        case Matches(env) => Some(Some((pat, env, result)))
+        case NotProvable => Some(None)
+        case NoMatch => None
+      }
+    })).get // Totallity of matches should ensure this will always find something unless something has gone terribly wrong
+
+  def solveMatch(env: PatternEnv, result: NormalExpression) =
+    ((env.size - 1) to 0 by -1).toList.map(env.get(_).get) // If this exceptions then somehow we didn't get enough names in the env
+      .foldLeft(result) { case (ne, arg) => NormalExpression.App(ne, arg) }
+ 
+  def normalOrderReduction(expr: NormalExpression): NormalExpression = {
+    import NormalExpression._
+    val nextExpr = expr match {
+      case App(Lambda(nextExpr), arg) => {
+        applyLambdaSubstituion(nextExpr, Some(arg), 0)
+      }
+      case m@Match(struct@Struct(enum, args), branches) =>
+        findMatch(m) match {
+          case None => m
+          case Some((pat, env, result)) => solveMatch(env, result)
+        }
+      case Recursion(Lambda(innerExpr)) if(innerExpr.maxLambdaVar.map(_ < 0).getOrElse(true)) => {
+        applyLambdaSubstituion(innerExpr, None, 0)
+      }
+      case Lambda(App(innerExpr, LambdaVar(0))) => {
+        innerExpr
+      }
+      case _ => expr
+    }
+    val res = nextExpr match {
+      case al @ App(Lambda(_), _) => normalOrderReduction(al)
+      case App(fn, arg) =>
+        normalOrderReduction(fn) match {
+          case l @ Lambda(_) => normalOrderReduction(App(l, arg))
+          case nfn @ _       => App(nfn, normalOrderReduction(arg))
+        }
+      case extVar @ ExternalVar(_, _) => extVar
+      case Match(arg, branches) =>
+        val nextMatch = Match(normalOrderReduction(arg), branches)
+        findMatch(nextMatch) match {
+          case None => nextMatch
+          case Some(_) => normalOrderReduction(nextMatch)
+        }
+      case lv @ LambdaVar(_)  => lv
+      case Lambda(expr)       =>
+        normalOrderReduction(expr) match {
+          case a@App(innerExpr, LambdaVar(0)) => normalOrderReduction(Lambda(a))
```

Comment:
this looks like it can be an infinite loop. Can you comment why it can't be? e.g. if you have `a == App(_, LambdaVar(0))` why does normalOrderReduction(a) not return the same? Is this just hard wiring knowledge of eta conversion here?

Like, why can `normalOrderReduction` ever return something that could be eta converted? That seems like that shouldn't happen.

Reply:
It shouldn't infinite loop in that case because of
```scala
      case Lambda(App(innerExpr, LambdaVar(0))) => {
        innerExpr
      }
```
above which will unwrap to the `_`.

The second question is harder for me to answer. I'll try to come up with an example

Reply:
Here's an example
```
Lambda(App(
  ExternalVar(
    PackageName(NonEmptyList.fromList(List("Extern","Simple")).get),Identifier.Name("foo"
  ),
  App(Lambda(LambdaVar(0), LambdaVar(0)))
))
```
The tail will get beta'd to `LambdaVar(0)` so this will get eta'd down to just the `ExternVar`